### PR TITLE
RavenDB-10033 Better logging for the Rachis

### PR DESCRIPTION
**Pull request by karmeli87**
WIP

- Currently only the follower and the follower candidate are recorded.
- And Tcp Ping endpoint.
- Set linger of the tcp client to zero, so it will close the socket immediatly on dispose without trying to send what it still have.

### CODE REVIEW

File: src/Raven.Server/Documents/Handlers/Debugging/NodeDebugHandler.cs [hunk 1]
```diff
@@ -35,5 +42,98 @@ public Task ListRemoteConnections()
...
+                        write.WriteComma();
+                    }
+                    write.Flush();
+                }
+                write.WriteEndObject();
+                write.Flush();
+            }
+        }
+
+        private async Task<(string url, string res)> PingOnce(string url)
```
Comment (ayende) [inline src/Raven.Server/Documents/Handlers/Debugging/NodeDebugHandler.cs]: I don't like this.
A) use PascalCase for tuple args
B) Use proper structure for the reply, not comma separated string, it wouldn't be readable in a year.
C) `Task<(string Url, (int TcpInfoTime, int SendTime, in RecieveTime), string Error)>` would do.

File: src/Raven.Server/Rachis/Follower.cs [hunk 2]
```diff
@@ -65,7 +67,7 @@ private void FollowerSteadyState()
 
                         return;
                     }
-
+                    _debugRecorder.Record($"Got {appendEntries.EntriesCount} entries");
```
Comment (ayende) [inline src/Raven.Server/Rachis/Follower.cs]: This is going to be running all the time, and it is going to always be allocating strings that we'll most of the time never look at.

File: src/Raven.Server/Rachis/FollowerAmbassador.cs [hunk 3]
```diff
@@ -230,8 +243,11 @@ private unsafe void Run()
 #endif
                                     );
                                 }
+                                _debugRecorder.Record($"Start sending {entries.Count} entries");
```
Comment (ayende) [inline src/Raven.Server/Rachis/FollowerAmbassador.cs]: Going to be called _all the time_ , and very rarely looked at. Too expensive.

File: src/Raven.Server/Rachis/RachisConsensus.cs [hunk 4]
```diff
@@ -120,12 +121,126 @@ public unsafe List<BlittableJsonReaderObject> GetLogEntries(long first, Transact
         }
     }
 
+    public class RachisLogEntry : IDynamicJsonValueConvertible
+    {
+        public DateTime At = DateTime.UtcNow;
```
Comment (ayende) [inline src/Raven.Server/Rachis/RachisConsensus.cs:126]: Shouldn't we output this as well?

File: src/Raven.Server/Rachis/RachisConsensus.cs [hunk 5]
```diff
@@ -120,12 +121,126 @@ public unsafe List<BlittableJsonReaderObject> GetLogEntries(long first, Transact
...
+            {
+                [nameof(Message)] = Message,
+                [nameof(Ticks)] = Ticks
+            };
+        }
+    }
+
+    public class RachisTimings
+    {
+        public readonly List<RachisLogEntry> Timings = new List<RachisLogEntry>();
```
Comment (ayende) [inline src/Raven.Server/Rachis/RachisConsensus.cs]: This is accessed from multiple threads concurrently in an unsafe manner.

File: src/Raven.Server/Rachis/RachisConsensus.cs [hunk 6]
```diff
@@ -120,12 +121,126 @@ public unsafe List<BlittableJsonReaderObject> GetLogEntries(long first, Transact
...
+
+        public DynamicJsonValue ToJson()
+        {
+            var timingTracking = new DynamicJsonValue();
+            foreach (var tuple in TimingTracking.ForceEnumerateInThreadSafeManner().OrderBy(x => x.Key))
+            {
+                var key = tuple.Key;
+                DynamicJsonArray inner;
+                timingTracking[key] = inner = new DynamicJsonArray();
+                using (var queue = tuple.Value.GetEnumerator())
```
Comment (ayende) [inline src/Raven.Server/Rachis/RachisConsensus.cs]: It's safe to iterate here with `foreach`

File: src/Raven.Server/Rachis/RachisConsensus.cs [hunk 7]
```diff
@@ -120,12 +121,126 @@ public unsafe List<BlittableJsonReaderObject> GetLogEntries(long first, Transact
...
+                {
+                    while (queue.MoveNext())
+                    {
+                        inner.Add(new DynamicJsonArray(queue.Current.Timings.OrderBy(x => x.At)));
+                    }  
+                }
+            }
+            
+            var stateTracking = new DynamicJsonArray();            
+            using (var queue = StateChangeTracking.GetEnumerator())
```
Comment (ayende) [inline src/Raven.Server/Rachis/RachisConsensus.cs]: Can use `foreach` here

File: src/Raven.Server/Rachis/RachisConsensus.cs [hunk 8]
```diff
@@ -120,12 +121,126 @@ public unsafe List<BlittableJsonReaderObject> GetLogEntries(long first, Transact
...
+                Message = message,
+                Ticks = _sp.ElapsedMilliseconds
+            });
+        }
+    }
+
+    public class RachisDebug
+    {
+        public readonly ConcurrentDictionary<string, ConcurrentQueue<RachisTimings>> TimingTracking = new ConcurrentDictionary<string, ConcurrentQueue<RachisTimings>>();
+        public readonly ConcurrentQueue<string> StateChangeTracking = new ConcurrentQueue<string>();
```
Comment (ayende) [inline src/Raven.Server/Rachis/RachisConsensus.cs:182]: What about _when_ something happened?

File: src/Raven.Server/ServerWide/ServerStore.cs [hunk 9]
```diff
@@ -483,13 +485,10 @@ private void OnStateChanged(object sender, RachisConsensus.StateTransition state
...
 
                 if (state.To == RachisState.LeaderElect  || state.To == RachisState.Leader)
-                {
-                    _engine.CurrentLeader.OnNodeStatusChange += OnTopologyChanged;
-                } else if (state.From == RachisState.LeaderElect || state.From == RachisState.Leader)
                 {
                     _engine.CurrentLeader.OnNodeStatusChange -= OnTopologyChanged;
-                }
-
+                    _engine.CurrentLeader.OnNodeStatusChange += OnTopologyChanged;
```
Comment (ayende) [inline src/Raven.Server/ServerWide/ServerStore.cs]: huh? You unregister and register again immediately?

File: src/Raven.Server/ServerWide/ServerStore.cs [hunk 10]
```diff
@@ -1202,7 +1201,7 @@ public void IdleOperations(object state)
             {
                 try
                 {
-                    _timer.Change(_frequencyToCheckForIdleDatabases, TimeSpan.FromDays(7));
+                    _timer.Change(_frequencyToCheckForIdleDatabases, _frequencyToCheckForIdleDatabases);
```
Comment (ayende) [inline src/Raven.Server/ServerWide/ServerStore.cs]: That was intentionally very high in the future, we don't want to have slow check that will become re-enterant.

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Server/ServerWide/ServerStore.cs (hunk 9 above)
Comment (karmeli87) [reply src/Raven.Server/ServerWide/ServerStore.cs]: To ensure that we have registered the event only once.

File: src/Raven.Server/Rachis/RachisConsensus.cs [hunk 11]
```diff
@@ -207,24 +207,14 @@ public DynamicJsonValue ToJson()
...
                 DynamicJsonArray inner;
                 timingTracking[key] = inner = new DynamicJsonArray();
-                using (var queue = tuple.Value.GetEnumerator())
+                foreach (var queue in tuple.Value)
                 {
-                    while (queue.MoveNext())
-                    {
-                        inner.Add(new DynamicJsonArray(queue.Current.Timings.OrderBy(x => x.At)));
-                    }  
+                    inner.Add(new DynamicJsonArray(queue.Timings.OrderBy(x => x.At)));
```
Comment (ayende) [inline src/Raven.Server/Rachis/RachisConsensus.cs]: Potential concurrency issue here (talk with @maximburyak )

**Pull request merged**